File: flink-app/src/main/java/com/example/ctr/config/KafkaProperties.java
package com.example.ctr.config;

import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.stereotype.Component;

import jakarta.validation.constraints.NotBlank;
import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
@Component
@ConfigurationProperties(prefix = "kafka")
public class KafkaProperties {

    @NotBlank
    private String bootstrapServers;

    @NotBlank
    private String groupId;

    private Topics topics = new Topics();

    @Getter
    @Setter
    public static class Topics {
        @NotBlank
        private String impression;
        @NotBlank
        private String click;
    }
}

File: flink-app/src/main/kotlin/com/example/ctr/config/KafkaProperties.kt
package com.example.ctr.config

import jakarta.validation.constraints.NotBlank
import jakarta.validation.Valid

class KafkaProperties {
    @field:NotBlank
    var bootstrapServers: String = ""

    @field:NotBlank
    var groupId: String = ""

    @field:NotBlank
    var offsetStrategy: String = "committed"

    @field:NotBlank
    var dlqTopic: String = "dlq-events"

    @field:Valid
    var topics: Topics = Topics()

    class Topics {
        @field:NotBlank
        var impression: String = ""

        @field:NotBlank
        var click: String = ""
    }

    companion object {
        private val VALID_OFFSET_STRATEGIES = setOf("latest", "earliest", "committed")

        fun validateOffsetStrategy(strategy: String): Boolean {
            return strategy.lowercase() in VALID_OFFSET_STRATEGIES
        }
    }

    fun toProducerProperties(): java.util.Properties {
        return java.util.Properties().apply {
            put("bootstrap.servers", bootstrapServers)
            put("group.id", groupId)
        }
    }
}
